database/database.go

```diff
@@ -449,6 +441,33 @@ func (db *DB) updateFailedFlag(hashes []Hash, failed bool) error {
 	return err
 }
 
+// ignoreDuplicateKeyErrors takes an error, if that error is a mongo
+// BulkWriteException, it will loop through the write errors and ignore
+// duplicate key errors. If all write errors were duplicate key errors, this
+// function returns nil, otherwise it simply returns the given error.
+func ignoreDuplicateKeyErrors(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	// check whether the given error is a BulkWriteException, if it's not simply
+	// return the error
+	bWriteErr, ok := err.(mongo.BulkWriteException)
+	if !ok {
+		return err
+	}
+
+	// loop over all write errors, ignore the duplicate key errors, if all write
+	// errors are duplicate key errors we want to ignore the bulk write error
+	// all together
+	for _, bWriteError := range bWriteErr.WriteErrors {
+		if !isDuplicateKey(bWriteError) {
+			return err
+		}
+	}
```

Comment:
While this does ignore all `DuplicateKey` errors, it also truncates the list of errors to the first one it encounters.
That is almost OK because any non-dup error will result in the current portal sync being aborted. But it might not give us all errors that occurred. Is that important? If not, feel free to resolve this.

Reply:
It actually simply returns the original error as soon as it finds a write error that's not a duplicate key error. But even if it weren't it's probably okay. This is an edge case that I don't expect to occur ever, and even if it does it's still not much of an issue.